src/Microsoft.ML.Data/Transforms/KeyToValueTransform.cs

```diff
@@ -67,386 +79,547 @@ private static VersionInfo GetVersionInfo()
+                Host.Check(keyMetadata.Length == typeKey.ItemType.KeyCount);
+
+                VBufferUtils.Densify(ref keyMetadata);
+                return new KeyToValueMap<TKey, TValue>(this, typeKey.ItemType.AsKey, typeVal.ItemType.AsPrimitive, keyMetadata.Values, iinfo);
+            }
             /// <summary>
-            /// The parent transform.
+            /// A map is an object capable of creating the association from an input type, to an output
+            /// type. This mapping is constructed from key metadata, with the input type being the key type
+            /// and the output type being the type specified by the key metadata.
             /// </summary>
-            protected readonly KeyToValueTransform Parent;
-
-            protected KeyToValueMap(KeyToValueTransform trans, PrimitiveType typeVal, int iinfo)
+            private abstract class KeyToValueMap
             {
-                // REVIEW: Is there a better way to perform this first assert value?
-                Contracts.AssertValue(trans);
-                Parent = trans;
-                Parent.Host.AssertValue(typeVal);
-                Parent.Host.Assert(0 <= iinfo && iinfo < Parent.Infos.Length);
-                TypeOutput = typeVal;
-                InfoIndex = iinfo;
-            }
-
-            public abstract Delegate GetMappingGetter(IRow input);
+                /// <summary>
+                /// The item type of the output type, that is, either the output type or,
+                /// if a vector, the item type of that type.
+                /// </summary>
+                protected readonly PrimitiveType TypeOutput;
+
+                /// <summary>
+                /// The column index in Infos.
+                /// </summary>
+                protected readonly int InfoIndex;
+
+                /// <summary>
+                /// The parent transform.
+                /// </summary>
+                protected readonly Mapper Parent;
+
+                protected KeyToValueMap(Mapper mapper, PrimitiveType typeVal, int iinfo)
+                {
+                    // REVIEW: Is there a better way to perform this first assert value?
+                    Contracts.AssertValue(mapper);
+                    Parent = mapper;
+                    Parent.Host.AssertValue(typeVal);
+                    Parent.Host.Assert(0 <= iinfo && iinfo < Parent._types.Length);
+                    TypeOutput = typeVal;
+                    InfoIndex = iinfo;
+                }
 
-            public abstract JToken SavePfa(BoundPfaContext ctx, JToken srcToken);
-        }
+                public abstract Delegate GetMappingGetter(IRow input);
 
-        private class KeyToValueMap<TKey, TValue> : KeyToValueMap
-        {
-            private readonly TValue[] _values;
-            private readonly TValue _na;
+                public abstract JToken SavePfa(BoundPfaContext ctx, JToken srcToken);
```

Comment:
>SavePfa [](start = 39, length = 7)

It's weird what it can save itself as pfa, but doesn't implement ISaveAsPfa interface.

Reply:
good catch

---
In reply to: [216510823](https://github.com/dotnet/machinelearning/pull/856#discussion_r216510823) [](ancestors = 216510823)